### ISSUE TITLE
Multiple JSDoc Annotations on Same Line Ignored

### ISSUE THREAD

**Issue by ChadKillingsworth** [Polymer/polymer-analyzer#475]
Closure compiler supports the following syntax - but the analyzer does not:

```js
/** @const @polymerBehavior */
```

**Comment by justinfagnani** [comment]
We doctrine for jsdoc parsing. I'll have to see why it's doing this...

**Comment by arthurevans** [comment]
That's actually following the JSDoc standard:

"Block tags always begin with an at sign (@). **Each block tag must be followed by a line break**, with the exception of the last block tag in a JSDoc comment."

(Emphasis added.)

The Google JS style guide allows for this, but it's not valid JSDoc.

**Comment by justinfagnani** [comment]
Thanks @arthurevans !

**Comment by ChadKillingsworth** [comment]
Closure-compiler used to enforce the same. However, it was found inside Google that engineers were frequently placing multiple annotations on the same line. As they were silently ignored, and didn't produce compiler warnings, this led to bugs. So the compiler was changed to support multiple per line.

In my case, I'm using Closure-compiler to bundle the code before passing it to the analyzer (my behaviors are in CommonJS modules). The compiler rewrites the modules into the global scope so that they can be analyzed. I'm currently having to do an additional step to ensure that only one JSDoc annotation is listed per line.

**Comment by justinfagnani** [comment]
We're using eslint's jsdoc parser, I'll have to check if they have an option to support multiple tags per line.

Passing compiler output to the Analyzer can other problems though. One we saw was compilers putting expressions under a comma operator to unset the receiver. This moves the jsdoc comments to the wrong node. Moving string literals around could easily be problematic too.

**Comment by ChadKillingsworth** [comment]
> Passing compiler output to the Analyzer can other problems though. One we saw was compilers putting expressions under a comma operator to unset the receiver. This moves the jsdoc comments to the wrong node. Moving string literals around could easily be problematic too.

I would expect this with both `SIMPLE` and `ADVANCED` modes. I'm using the badly named `WHITESPACE_ONLY` mode. It's sole purpose is to rewrite the module references so the analyzer doesn't have to understand CommonJS modules.

**Comment by justinfagnani** [comment]
I didn't know anyone was using CommonJS with the Analyzer yet, but it wouldn't be difficult to support. If the Analyzer understands the original source you'll get better IDE/language service support.

We do have a JS module import scanner, is just not wired in. Could you use that?

With either import syntax, we would need a node URL resolver.

**Comment by ChadKillingsworth** [comment]
I just wrote the node module resolution algorithm for the compiler :)  In my case, all the relevant modules for the compiler are proper path references (i.e. the require paths all start with either '/', '../' or './'). I'm using commonJS because it doesn't require transpilation. The InjectJS loader from Linked In is old, but allows you to use CommonJS in the browser (this is for development mode - the files are all bundled by closure-compiler for production).

I could definitely use an import scanner like that. I would much prefer that to having to pre-bundle code for the analyzer. I just needed to get something working so I could move on ...

**Comment by justinfagnani** [comment]
I filed https://github.com/eslint/doctrine/issues/191 with doctrine for an option to support multiple tags per line, so let's see how that's received. Other options are to look for another parser, fork, or preprocess the comments to insert newlines.

**Comment by justinfagnani** [comment]
@ChadKillingsworth if you're using the Analyzer with CJS, could you file an issue for optional support?

**Comment by ChadKillingsworth** [comment]
@justinfagnani FYI I'm working on a change that for at least the polymer specific annotations they will be on their own line post Closure-compiler. I can't do anything if an author places them both on the same line though.
